examples/fp_delegation.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unifex/for_each.hpp>
+#include <unifex/manual_event_loop.hpp>
+#include <unifex/range_stream.hpp>
+#include <unifex/scheduler_concepts.hpp>
+#include <unifex/sync_wait.hpp>
+#include <unifex/timed_single_thread_context.hpp>
+#include <unifex/transform_stream.hpp>
+#include <unifex/via_stream.hpp>
+#include <unifex/with_query_value.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <variant>
+
+using namespace unifex;
+using namespace std::chrono_literals;
+
+namespace {
+
+class delegating_scheduler;
+
+class delegating_context {
+  public:
+  delegating_context(int capacity) : capacity_{capacity} {}
+
+  bool reserve() {
+    auto lck = std::scoped_lock{m_};
+    bool reservation_made = reservation_count_ < capacity_;
+    if(reservation_made) {
+      ++reservation_count_;
+    }
+    return reservation_made;
+  }
+
+  int get_count() {
+    return run_count_;
+  }
+
+  void run() {
+    run_count_++;
+  }
+
+  delegating_scheduler get_scheduler() noexcept;
+
+  private:
+  friend class delegating_sender;
+  std::mutex m_;
+  int reservation_count_{0};
+  std::atomic<int> run_count_{0};
+  int capacity_;
+  timed_single_thread_context single_thread_context_;
+};
+
+template <typename DelegatedOperationState, typename LocalOperationState>
+class delegating_operation final {
+  public:
+  delegating_operation(DelegatedOperationState&& op, delegating_context* context) : op_{std::move(op)}, context_{context} {}
+  delegating_operation(LocalOperationState&& op, delegating_context* context) : op_{std::move(op)}, context_{context} {}
+
+  inline void start() noexcept {
+    if(std::holds_alternative<DelegatedOperationState>(op_)) {
+      // Start a delegated operation
+      std::get<DelegatedOperationState>(op_).start();
+    } else {
+      // Start immediately on the local context
+      context_->run();
+      std::get<LocalOperationState>(op_).start();;
+    }
+  }
+
+  std::variant<DelegatedOperationState, LocalOperationState> op_;
+  delegating_context* context_ = nullptr;
+};
+
+class delegating_sender {
```

Comment:
This kind of variant-sender could probably be factored out into a separate `variant_sender` helper class that implements the more general "this could be one of several senders" patterns.

Reply:
Yes. Maybe not for this example, though. It could be a useful primitive, and if we decide a first step is to make the delegating context a core unifex piece rather than just in this example.